### PR TITLE
boot: add boot.Modeenv that can read/write the UC20 modeenv files

### PR DESCRIPTION
**Pull request by mvo5**
This commits adds basic support for the modeenv file that will be
used on UC20 systems to keep track of the run mode, the recovery
system to use (and more later).

### CODE REVIEW

File: boot/modeenv.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+func (m *Modeenv) Write(rootdir string) error {
+	modeenvPath := filepath.Join(rootdir, dirs.SnapModeenvFile)
+
+	if err := os.MkdirAll(filepath.Dir(modeenvPath), 0755); err != nil {
+		return err
+	}
+	modeEnvContent := fmt.Sprintf(`recovery_system=%s
+mode=%s
+`, m.RecoverySystemLabel, m.Mode)
+	if err := ioutil.WriteFile(modeenvPath, []byte(modeEnvContent), 0644); err != nil {
```
Comment (pedronis) [inline boot/modeenv.go]: should we use an atomic write for this?

File: boot/modeenv.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/mvo5/goconfigparser"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+// Modeenv is a file on UC20 that provides additional information
+// about the current mode (run,recover,install)
+type Modeenv struct {
+	Mode                string
+	RecoverySystemLabel string
```
Comment (pedronis) [inline boot/modeenv.go]: s/RecoverSystemLabel/RecoverySystem/ perhaps?

File: boot/modeenv.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (m *Modeenv) Write(rootdir string) error {
+	modeenvPath := filepath.Join(rootdir, dirs.SnapModeenvFile)
+
+	if err := os.MkdirAll(filepath.Dir(modeenvPath), 0755); err != nil {
+		return err
+	}
+	modeEnvContent := fmt.Sprintf(`recovery_system=%s
+mode=%s
```
Comment (pedronis) [inline boot/modeenv.go]: we'll need to add base as well later etc. Maybe mode should go first.

File: boot/modeenv_test.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func (s *modeenvSuite) TestWriteExisting(c *C) {
+	s.makeMockModeenvFile(c, "mode: run")
+
+	modeenv, err := boot.ReadModeenv(s.tmpdir)
+	c.Assert(err, IsNil)
+	modeenv.Mode = "recovery"
+	err = modeenv.Write(s.tmpdir)
+	c.Assert(err, IsNil)
```
Comment (pedronis) [inline boot/modeenv_test.go]: we should read it again, no?

Comment (pedronis) [reply boot/modeenv_test.go]: also we should have a write non-existing test?

File: boot/modeenv_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	s.makeMockModeenvFile(c, "mode: run")
+
+	modeenv, err := boot.ReadModeenv(s.tmpdir)
+	c.Assert(err, IsNil)
+	c.Check(modeenv.Mode, Equals, "run")
+	c.Check(modeenv.RecoverySystem, Equals, "")
+}
+
+func (s *modeenvSuite) TestReadModeWithRecoverySystem(c *C) {
+	s.makeMockModeenvFile(c, `mode: recovery
```
Comment (pedronis) [inline boot/modeenv_test.go]: should we use = to be consistent with how we write?

Comment (mvo5) [reply boot/modeenv_test.go]: Yes, silly me!

**Pull request merged**